tests/unit/iontests.js

```diff
@@ -59,24 +64,130 @@ define([
         }
 
         function exhaust(reader) {
-            for (;;) {
-                let next = reader.next();
-                if (typeof(next) === 'undefined') {//should this be null?
-                    if (reader.depth() > 0) {
-                        // End of container
-                        reader.stepOut();
-                    } else {
-                        // End of data
-                        break;
-                    }
-                } else if (next.container && !reader.isNull()) {
+            for (let type; type = reader.next(); ) {
+                if (type.container && !reader.isNull()) {
                     reader.stepIn();
+                    exhaust(reader);
+                    reader.stepOut();
                 } else {
                     reader.value();
                 }
             }
         }
 
+        function readerCompareTest(source) {
+            function toBytes(source, writer) {
+                let reader = ion.makeReader(source);
+                writeTo(reader, writer);
+                writer.close();
+                return writer.getBytes();
+            }
+
+            let ionBinary = toBytes(source, ion.makeBinaryWriter());
+            let ionText = toBytes(source, ion.makeTextWriter());
+
+            readerCompare(ion.makeReader(ionBinary), ion.makeReader(ionText));
+        }
+
+        function readerCompare(r1, r2) {
+            checkReaderValueMethods(r1, r2, null);
+
+            while (true) {
+                let t1 = r1.next();
+                let t2 = r2.next();
+
+                let v;
+                try {
+                    v = r1.value();
+                } catch (e) {
+                    v = 'value() threw ' + e.message;
+                }
+                assert.equal(t1, t2);
+
+                checkReaderValueMethods(r1, r2, t1);
+
+                if (t1 === null) {
+                    break;
+                }
+
+                if (t1.container && !r1.isNull()) {
+                    r1.stepIn();
+                    r2.stepIn();
+                    readerCompare(r1, r2);
+                    r1.stepOut();
+                    r2.stepOut();
+                }
+            }
+        }
+
+        function checkReaderValueMethods(r1, r2, type) {
+            allValueMethods.forEach(methodName => {
+                let typeName = type == null ? 'null' : type.name;
+                if (type != null && nonThrowingMethods[typeName][methodName]) {
+                    let v1 = r1[methodName]();
+                    let v2 = r2[methodName]();
+                    assert(v1 !== undefined, "unexpected 'undefined' response");
+                    assert(v2 !== undefined, "unexpected 'undefined' response");
+                    assert.deepEqual(v1, v2, methodName + '():  ' + v1 + ' != ' + v2);
+                } else {
+                    assert.throws(() => { r1[methodName]() }, '', '', 'Expected ' + methodName + '() to throw');
```

Comment:
How do you know this isn't throwing for some reason other than the one you intend?

Reply:
We don't at this time; we could check for specific error messages, but that would be brittle and unwarranted, imo.  Based on a quick search, JavaScript support for custom errors appears to be minimal, and I haven't seen any code that does so.

Perhaps TypeScript includes capabilities that would be of help here.  I've opened #292 to explore this further.